### PR TITLE
smart-pager increment pager correctly (according to museumNumber)

### PR DESCRIPTION
**Pull request by yCobanoglu**

### CODE REVIEW

File: ebl/fragmentarium/infrastructure/fragment_repository.py [hunk 1]
```diff
@@ -261,34 +264,109 @@ def get_numbers(pipeline):
...
-            self._fragments.find_many({"_id": {"$gt": f"{str(number)}"}})
-            .sort("_id", 1)
-            .limit(1)
-        )
-        previous = (
-            self._fragments.find_many({"_id": {"$lt": f"{str(number)}"}})
-            .sort("_id", -1)
-            .limit(1)
-        )
+    def query_next_and_previous_fragment(
```
[the remark this answers is not in the document]
Comment (jlaasonen) [reply ebl/fragmentarium/infrastructure/fragment_repository.py:267]: This must be fixed the code is very messy. It is hard to figure out what is going on. The logic appears to be a duplicate of the `MuseumNumber` comparison. If the existing logic cannot be used, the common parts should be extracted.

File: ebl/tests/fragmentarium/test_fragment_repository.py [hunk 2]
```diff
@@ -158,32 +158,54 @@ def test_find_random(fragment_repository):
...
+    "CBS.0",
+    "UM.0",
+    "N.0",
+    "N.1",
+    "1841-57.54",
+    "1841-57.57",
+    "1841-57.63",
+    "X.0",
+    "X.1",
+]
```
Comment (jlaasonen) [inline ebl/tests/fragmentarium/test_fragment_repository.py:181]: There should be test cases with suffixes.

File: ebl/fragmentarium/infrastructure/fragment_repository.py [hunk 3]
```diff
@@ -261,34 +264,109 @@ def get_numbers(pipeline):
...
+                else:
+                    return adjacent_museum_number
             raise NotFoundError("Could not retrieve any fragments")
-        else:
-            return result
+
+        return {
+            "previous": iterate_until_found(prev, -1),
+            "next": iterate_until_found(next, 1),
+        }
```
Comment (jlaasonen) [inline ebl/fragmentarium/infrastructure/fragment_repository.py:369]: Use a proper model instead of passing dictionaries around.

**Review by jlaasonen — CHANGES_REQUESTED** [review]

File: ebl/tests/fragmentarium/test_fragment_repository.py (hunk 2 above)
Comment (jlaasonen) [reply ebl/tests/fragmentarium/test_fragment_repository.py:181]: ..and non-numeric number.

Comment (yCobanoglu) [reply ebl/tests/fragmentarium/test_fragment_repository.py:181]: "1841-57" is non numeric or isn't that what you meant ?

Comment (jlaasonen) [reply ebl/tests/fragmentarium/test_fragment_repository.py]: Non-numeric **number**. "1841-57" is a **prefix** in the data above.

E.g "Fragment" in "Test.Fragment"

**Review by jlaasonen — CHANGES_REQUESTED** [review]
The complexity issues must be fixed. The code is too hard to understand.

**Comment by yCobanoglu** [comment]
[Use parsed museum number for smart pager](https://trello.com/c/VKmEfqYY/397-use-parsed-museum-number-for-smart-pager)

File: ebl/fragmentarium/infrastructure/fragment_repository.py [hunk 4]
```diff
@@ -261,34 +263,83 @@ def get_numbers(pipeline):
...
-        next_ = (
-            self._fragments.find_many({"_id": {"$gt": f"{str(number)}"}})
-            .sort("_id", 1)
-            .limit(1)
+    def _compare_within_two_museum_numbers(
+        self,
+        museum_number: MuseumNumber,
+        current_museum_number: MuseumNumber,
+        current_prev_or_next: Optional[MuseumNumber],
+        comp: Callable[[MuseumNumber, MuseumNumber], bool],
```
Comment (jlaasonen) [inline ebl/fragmentarium/infrastructure/fragment_repository.py:271]: What's a "comp"? Do not unnecessarily truncate words.

File: ebl/fragmentarium/infrastructure/fragment_repository.py [hunk 5]
```diff
@@ -261,34 +263,83 @@ def get_numbers(pipeline):
...
+    ) -> Tuple[Optional[MuseumNumber], Optional[MuseumNumber]]:
+        if first is None or current_museum_number < first:
+            first = current_museum_number
+        if last is None or current_museum_number > last:
+            last = current_museum_number
+        return first, last
+
+    def _find_adjacent_museum_number_from_list(
+        self, museum_number: MuseumNumber, cursor, is_endpoint=False
+    ):
```
Comment (jlaasonen) [inline ebl/fragmentarium/infrastructure/fragment_repository.py:295]: Add type hints for `cursor` and return value.

Use a descriptive name. There is no list involved.

File: ebl/fragmentarium/infrastructure/fragment_repository.py [hunk 6]
```diff
@@ -261,34 +263,83 @@ def get_numbers(pipeline):
...
+        if is_endpoint:
+            current_prev = current_prev or last
+            current_next = current_next or first
+        return current_prev, current_next
+
+    def query_next_and_previous_fragment(
+        self, museum_number: MuseumNumber
+    ) -> FragmentPagerInfo:
+
+        self.query_by_museum_number(museum_number)
```
Comment (jlaasonen) [inline ebl/fragmentarium/infrastructure/fragment_repository.py:325]: What's this for? The fragment is not used.

File: ebl/fragmentarium/infrastructure/fragment_repository.py [hunk 7]
```diff
@@ -261,34 +263,83 @@ def get_numbers(pipeline):
...
+        self,
+        first: Optional[MuseumNumber],
+        last: Optional[MuseumNumber],
+        current_museum_number: MuseumNumber,
+    ) -> Tuple[Optional[MuseumNumber], Optional[MuseumNumber]]:
+        if first is None or current_museum_number < first:
+            first = current_museum_number
+        if last is None or current_museum_number > last:
+            last = current_museum_number
+        return first, last
```
Comment (jlaasonen) [inline ebl/fragmentarium/infrastructure/fragment_repository.py:291]: `self` is not used. These could be moved out of the class.

The names are misleading. Compare would indicate that the result tells if something is greater/less than or equal.

**Review by jlaasonen — CHANGES_REQUESTED** [review]

**Review by jlaasonen — APPROVED** [review]

File: ebl/fragmentarium/infrastructure/fragment_repository.py (hunk 6 above)
Comment (yCobanoglu) [reply ebl/fragmentarium/infrastructure/fragment_repository.py:325]: this is used to check if the museum number exists at all

**Pull request merged**